### ISSUE TITLE
Request hangs forever with ClientRequest

### ISSUE THREAD

**Issue by 95th** [mswjs/interceptors#259]
I am facing an issue using msw where the request is responded with `node-fetch` but hangs forever with `ClientRequest`.
The last thing logged to console with `DEBUG=*` is the following:

```
  http:request POST https://localhost:8080/token chunk successfully stored! [
  <Buffer 67 72 61 6e 74 5f 74 79 70 65 3d 70 61 73 73 77 6f 72 64 26 73 63 6f 70 65 3d 41 64 6f 62 65 49 44 25 32 43 63 72 65 61 74 69 76 65 5f 73 64 6b 26 63 ... 128 more bytes>
] +10s
  http:request POST https://localhost:8080/token event:socket +19s
```

~~Could you please provide some guidance on how to debug this further?~~

I made some progress in debugging. My library uses `https` from `follow-redirect` which has the following `end` impl:

```js
// Ends the current native request
RedirectableRequest.prototype.end = function (data, encoding, callback) {
  // Shift parameters if necessary
  if (typeof data === "function") {
    callback = data;
    data = encoding = null;
  }
  else if (typeof encoding === "function") {
    callback = encoding;
    encoding = null;
  }

  // Write data if needed and end
  if (!data) {
    this._ended = this._ending = true;
    this._currentRequest.end(null, null, callback);
  }
  else {
    var self = this;
    var currentRequest = this._currentRequest;
    this.write(data, encoding, function () {
      self._ended = true;
      currentRequest.end(null, null, callback);
    });
    this._ending = true;
  }
};
```

In case of mocking with msw, the callback in else is never called but it is called in real calls.

```js
function () {
  self._ended = true;
  currentRequest.end(null, null, callback);
}
```

So I am guessing there is some issue with handling this case?

Ok I have a test case for this:

```js
test('preserves original POST request JSON body', async () => {
  const data = JSON.stringify({ todo: 'Buy the milk' })

  const options = {
    hostname: 'localhost',
    port: 443,
    path: '/',
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': data.length,
    },
  }

  const req = https.request(options, (res) => {
    console.log(`statusCode: ${res.statusCode}`)

    res.on('data', (d) => {
      process.stdout.write(d)
    })
  })

  req.on('error', (error) => {
    console.error(error)
  })

  req.end(data)  //  this doesnt work

  expect(requests).toHaveLength(1)  // fails. Length is 0
  const [request] = requests
  expect(request.method).toBe('POST')
  expect(request.body).toEqual(JSON.stringify({ todo: 'Buy the milk' }))
})
```
if I change the `req.end(data)` to below, then test case passes:
```js
req.write(data)
req.end()
```

**Comment by 95th** [comment]
It seems like `end` is never called with ClientRequest.

**Comment by 95th** [comment]
Also note that the same library is working correct for real endpoints. It just hanging in mocked case.

**Comment by 95th** [comment]
The library makes the following call with 178 bytes of data:
```js
req.end(data);
```

This trigger only `write` in the interceptor but not the `end`.

**Comment by 95th** [comment]
The fix seems to be calling the callback immediately. Since the writing cannot fail ever - we skip storing the callback:

```js
  write(...args: ClientRequestWriteArgs): boolean {
    const [chunk, encoding, callback] = normalizeClientRequestWriteArgs(args)
    this.log('write:', { chunk, encoding, callback })
    this.chunks.push({ chunk, encoding })
    this.requestBody = concatChunkToBuffer(chunk, this.requestBody)
    this.log('chunk successfully stored!', this.requestBody)

    /**
     * Prevent invoking the callback if the written chunk is empty.
     * @see https://nodejs.org/api/http.html#requestwritechunk-encoding-callback
     */
    if (!chunk || chunk.length === 0) {
      this.log('written chunk is empty, skipping callback...')
    } else {
      callback?.()
    }

    // Do not write the request body chunks to prevent
    // the Socket from sending data to a potentially existing
    // server when there is a mocked response defined.
    return true
  }
```

@kettanaito WDYT?

**Comment by 95th** [comment]
Oh wait.. that makes the _deciding later_ part complicated?

**Comment by 95th** [comment]
>  Since the writing cannot fail ever - we skip storing the callback:

it can if there is no mocked response and we have to go to the real endpoint?

**Comment by 95th** [comment]
I can confirm that the fix works for my case.

**Comment by kettanaito** [comment]
Hey, @95th. Thank you for reporting and looking into this! 

I've been tackling an issue with `follow-redirects` (https://github.com/mswjs/msw/issues/1125), perhaps it's connected to this one. I will review your pull request as soon as I have a minute. Oh, that'd be so fantastic if your changes would also solve the other issue there is.

**Comment by 95th** [comment]
@kettanaito I went through the other issue and it looks like it is the same issue.

**Comment by 95th** [comment]
Also, I looked into your issue on follow redirect and I agree with https://github.com/follow-redirects/follow-redirects/issues/195#issuecomment-1134588319

We're buffering the chunks. Buffering shouldn't provide guarantees at chunk level but rather at the flushing point which is the `end` call. if the end call has any error, it should invoke end's callback with error.

**Comment by kettanaito** [comment]
## Released: v0.16.5 🎉

This has been released in v0.16.5!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.16.5)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.16.5)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
